include/sympack/symPACKMatrix2D.hpp

```diff
@@ -1005,26 +1228,105 @@ namespace symPACK{
+                  tgt_width * src_snode_size > symPACK::gemm_limit ||
+                  src_nrows * src_snode_size > symPACK::gemm_limit) {
+                dev_ptr d_pivot_nzval = symPACK::gpu_allocator.allocate<T>(tgt_width * src_snode_size);
+                dev_ptr d_facing_nzval = symPACK::gpu_allocator.allocate<T>(src_nrows * src_snode_size);
+                dev_ptr d_buf = symPACK::gpu_allocator.allocate<T>(ldbuf * src_nrows);
+                if (d_pivot_nzval==nullptr) {
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else if (d_facing_nzval==nullptr) {
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else if (d_buf==nullptr) { 
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else {
+                    upcxx::when_all(
+                        upcxx::copy(pivot_nzval, d_pivot_nzval, tgt_width*src_snode_size),
+                        upcxx::copy(facing_nzval, d_facing_nzval, src_nrows*src_snode_size),
+                        upcxx::copy(buf, d_buf, ldbuf*src_nrows)
+                    ).wait();
+                    
+                    cublas::cublas_gemm_wrapper2(CUBLAS_OP_T, CUBLAS_OP_N,
+                              tgt_width, src_nrows, src_snode_size,
+                              T(-1.0), 
+                              symPACK::gpu_allocator.local(d_pivot_nzval), src_snode_size,
+                              symPACK::gpu_allocator.local(d_facing_nzval), src_snode_size, beta, 
+                              symPACK::gpu_allocator.local(d_buf), ldbuf);
+                    CUDA_ERROR_CHECK(cudaDeviceSynchronize());
+                    
+                    upcxx::copy(d_buf, buf, ldbuf*src_nrows).wait();
+                    
+                    symPACK::gpu_allocator.deallocate(d_pivot_nzval);
+                    symPACK::gpu_allocator.deallocate(d_facing_nzval);
+                    symPACK::gpu_allocator.deallocate(d_buf);
+                }
 
+              } else {
+                 blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                            T(-1.0),pivot_nzval,src_snode_size,
+                            facing_nzval,src_snode_size,beta,buf,ldbuf);
+              }
+#else
               blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
                   T(-1.0),pivot_nzval,src_snode_size,
                   facing_nzval,src_snode_size,beta,buf,ldbuf);
+#endif                  
```

Comment:
This is purely a stylistic observation, but I believe this pattern (which is repeated multiple times in this file) could more concisely be written as:

```suggestion
              } else 
#endif
              blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
                  T(-1.0),pivot_nzval,src_snode_size,
                  facing_nzval,src_snode_size,beta,buf,ldbuf);     
```

That way you have one fewer copy of the code making the complicated blas call, which helps with maintainability.

A possibly superior approach is to put the complicated blas call in a zero-argument lambda expression assigned to a variable at the entry to this logic, and then invoke that lambda for all the "use the CPU" cases (including the device allocation failure cases).

Reply:
Fixed in [97a3e5f](https://github.com/symPACK/symPACK/pull/12/commits/97a3e5f1f6d37fa0e56f7c6290ef8c9722114254)